pulp_rpm/app/serializers/advisory.py

```diff
@@ -125,52 +125,58 @@ def create(self, validated_data):
 
         update_collection_packages_to_save = list()
         update_references_to_save = list()
-        try:
-            update_record = super().create(validated_data)
-        except IntegrityError:
-            raise serializers.ValidationError("Advisory already exists in Pulp.")
-
-        for collection in pkglist:
-            new_coll = copy.deepcopy(collection)
-            packages = new_coll.pop("packages", [])
-            new_coll[PULP_UPDATE_COLLECTION_ATTRS.SHORTNAME] = new_coll.pop("short", "")
-            coll = UpdateCollection(**new_coll)
-            coll.update_record = update_record
-            coll.save()
-            for package in packages:
-                pkg = UpdateCollectionPackage(**package)
-                try:
-                    pkg.sum_type = createrepo_c.checksum_type(pkg.sum_type)
-                except TypeError:
-                    raise TypeError(f'"{pkg.sum_type}" is not supported.')
-                pkg.update_collection = coll
-                update_collection_packages_to_save.append(pkg)
-        for reference in references:
-            new_ref = dict()
-            new_ref[PULP_UPDATE_REFERENCE_ATTRS.HREF] = reference.get(
-                CR_UPDATE_REFERENCE_ATTRS.HREF, ""
-            )
-            new_ref[PULP_UPDATE_REFERENCE_ATTRS.ID] = reference.get(
-                CR_UPDATE_REFERENCE_ATTRS.ID, ""
-            )
-            new_ref[PULP_UPDATE_REFERENCE_ATTRS.TITLE] = reference.get(
-                CR_UPDATE_REFERENCE_ATTRS.TITLE, ""
-            )
-            new_ref[PULP_UPDATE_REFERENCE_ATTRS.TYPE] = reference.get(
-                CR_UPDATE_REFERENCE_ATTRS.TYPE, ""
-            )
-            ref = UpdateReference(**new_ref)
-            ref.update_record = update_record
-            update_references_to_save.append(ref)
-
-        if update_collection_packages_to_save:
-            UpdateCollectionPackage.objects.bulk_create(update_collection_packages_to_save)
-        if update_references_to_save:
-            UpdateReference.objects.bulk_create(update_references_to_save)
-
-        cr_update_record = update_record.to_createrepo_c()
-        update_record.digest = hash_update_record(cr_update_record)
-        update_record.save()
+        with transaction.atomic():
```

Comment:
I have a slight preference for keeping only the DB ops under the transaction, and leaving anything that won't throw exceptions outside.  Not enough to block if you disagree.

Reply:
Generally concur - but in this case, if **anything at all** goes wrong with the various pieces/parts of an advisory, we want to roll the entire effort back. In the course of this block we persist the advisory, collections, bulk-save packages and references, and then save the advisory again once we have a real digest. Putting the whole thing in a transaction is safer and clearer than trying to deal with "what if just this one thing failed?"

Reply:
I actually missed a `.save()` or two, so I'll just say "you're right" on this one